### ISSUE TITLE
Enable striping in rook

### ISSUE THREAD

**Issue by jomeier** [rook/rook#1960]
Hi,

we use rook in kubernetes clusters at Microsoft Azure.

Because of ceph is able to stripe data over several disks (which increases data throughput a lot) it would be great if rook also would be able to use striping.

At first it would be sufficient in my opinion if default ceph paramters are used.

For more info:
• https://stackoverflow.com/a/28698595
• http://programmability.us/mantle/blog3-striping#changing-the-striping-strategy
• http://docs.ceph.com/docs/jewel/man/8/rbd/#striping
• http://docs.ceph.com/docs/master/architecture/

Thanks and best regards,

Josef

**Comment by travisn** [comment]
@jomeier do you have more details on what settings would be needed to implement striping? Some of the settings have defaults that currently should be working in Rook. 

If a setting is needed in ceph.conf, a workaround is to apply the setting according to [these instructions](https://rook.io/docs/rook/v0.8/advanced-configuration.html#custom-cephconf-settings). For example, it looks like an rgw setting can be set in the ceph.conf, but others need flags passed. It would be great if you're able to test the config and give feedback on the improvements so we can prioritize the work accordingly.

**Comment by jomeier** [comment]
@travisn
The default stripe_count is set to one. Other stripe_counts require the kernel to have the stripingv2 (aka fancy stripe) feature which was introduced in Linux Kernel v4.17.

This are the parameters (description taken from the rbd documentation):

‘‘‘
The striping is controlled by three parameters:

object-size
The size of objects we stripe over is a power of two. It will be rounded up the nearest power of two. The default object size is 4 MB, smallest is 4K and maximum is 32M.

stripe_unit
Each [stripe_unit] contiguous bytes are stored adjacently in the same object, before we move on to the next object.

stripe_count
After we write [stripe_unit] bytes to [stripe_count] objects, we loop back to the initial object and write another stripe, until the object reaches its maximum size. At that point, we move on to the next [stripe_count] objects.

By default, [stripe_unit] is the same as the object size and [stripe_count] is 1. Specifying a different [stripe_unit] requires that the STRIPINGV2 feature be supported (added in Ceph v0.53) and format 2 images be used.
‘‘‘

I have to admit that I didn‘t completely understand if ceph stripes with a stripe_count of 1.

Is striping enabled in this case?

**Comment by ben-webber** [comment]
I too would like to be able to enable fancy striping for ceph using rook. According to this:

http://lists.ceph.com/pipermail/ceph-users-ceph.com/2018-January/024230.html

Since luminous, ceph now ignores the default rbd_default_stripe_count and rbd_default_stripe_unit that can be set in ceph.conf and requires that for striping to be enabled you need to specify stripe-count and stipe-unit as arguments to the rbd create command. I can confirm that setting these options when creating an image does enable fancy striping, but creating an image and omitting them does not, even when they are specified in the rook-config-override configmap:

```
ben@draco` ~ % oc --as=system:admin -n rook-ceph get configmap rook-config-override -o yaml
apiVersion: v1
data:
  config: |
    [global]
    public network = "10.9.20.0/24"
    cluster network = "10.9.21.0/24"
    public addr = ""
    cluster addr = ""
    [client]
    rbd default stripe_count = 16
    rbd default stripe_unit = 4194304
    rbd default features = 3
kind: ConfigMap
metadata:
  creationTimestamp: 2019-01-21T02:49:09Z
  name: rook-config-override
  namespace: rook-ceph
  resourceVersion: "128882"
  selfLink: /api/v1/namespaces/rook-ceph/configmaps/rook-config-override
  uid: 2069c198-1d27-11e9-8364-525400a0b9f8
ben@draco ~ % oc --as=system:admin -n rook-ceph exec -it rook-ceph-tools-6ddb786746-nhzq9 -- bash
[root@cerberus /]# rbd create replicapool/test --size=1G --image-feature striping,layering                                        
[root@cerberus /]# rbd -p replicapool info test
rbd image 'test':
        size 1 GiB in 256 objects
        order 22 (4 MiB objects)
        id: 14d966b8b4567
        block_name_prefix: rbd_data.14d966b8b4567
        format: 2
        features: layering
        op_features: 
        flags: 
        create_timestamp: Mon Jan 21 16:58:06 2019
[root@cerberus /]# rbd create replicapool/test1 --size=1G --image-feature striping,layering --stripe-count=16 --stripe-unit 4194304
[root@cerberus /]# rbd -p replicapool info test1
rbd image 'test1':
        size 1 GiB in 256 objects
        order 22 (4 MiB objects)
        id: 14d9c6b8b4567
        block_name_prefix: rbd_data.14d9c6b8b4567
        format: 2
        features: layering, striping
        op_features: 
        flags: 
        create_timestamp: Mon Jan 21 16:58:34 2019
        stripe unit: 4 MiB
        stripe count: 16
```

Looking at the following however, there doesn't appear to be a way to add additional arguments to the rbd create command when an image is created by rook:

https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/image.go#L101-L107

**Comment by mickaeldimonte-bib** [comment]
hello there,
would there be any updates on enabling ceph striping through rook?

**Comment by travisn** [comment]
It sounds like what's missing is the CSI driver would need to pick up the ceph.conf overrides to apply the settings during the rbd create. @Madhu-1 let's look at what it would take to get that working.

**Comment by kahirokunn** [comment]
Is there any progress? 👀

**Comment by leseb** [comment]
> Is there any progress? 👀

None, as mentioned in https://github.com/rook/rook/issues/1960#issuecomment-799552138 the fix is mostly in ceph-csi and a small piece in Rook. @Madhu-1 any inputs from the ceph-csi perspective?

**Comment by beegmon** [comment]
Bumping this as being able to set the object size, strip unit, and stripe count are use in our use-case (DBs on RDB images) and being able to better tune the underlying stripe count/unit/object size to something that better matches out DB page sizes may help improve overall performance for our workload.

**Comment by Greg-Tarr** [comment]
I would like this 👀

**Comment by travisn** [comment]
@Madhu-1 As mentioned in [this comment](https://github.com/rook/rook/issues/1960#issuecomment-456146217), can the csi driver create the rbd images with the parameters stripe-count and stipe-unit? If not, how about an issue be opened in the ceph-csi repo?

**Comment by Madhu-1** [comment]
@travisn @leseb not sure how I missed 3 notifications on the issue, My Apologies for missing notifications on the thread. if this can be mentioned in the ceph.conf https://rook.io/docs/rook/v1.9/ceph-advanced-configuration.html#custom-csi-cephconf-settings, if this needs to be done at the rbd image creation. As you mentioned I would suggest opening up an issue in https://github.com/ceph/ceph-csi repo.

**Comment by Madhu-1** [comment]
@idryomov can we expose stripe-count and stripe-unit in the RBD storageclass and use them for creating RBD images? is there any caveat here?

**Comment by idryomov** [comment]
> I too would like to be able to enable fancy striping for ceph using rook. According to this:
> 
> http://lists.ceph.com/pipermail/ceph-users-ceph.com/2018-January/024230.html
> 
> Since luminous, ceph now ignores the default rbd_default_stripe_count and rbd_default_stripe_unit that can be set in ceph.conf and requires that for striping to be enabled you need to specify stripe-count and stipe-unit as arguments to the rbd create command.

This is incorrect.  `rbd_default_stripe_unit` and `rbd_default_stripe_count` are respected:

```
$ ceph config set global rbd_default_stripe_unit 4096
$ ceph config set global rbd_default_stripe_count 4
$ rbd create --size 1G img
$ rbd info img
rbd image 'img':
        size 1 GiB in 256 objects
        order 22 (4 MiB objects)
        snapshot_count: 0
        id: 109b78c030ef
        block_name_prefix: rbd_data.109b78c030ef
        format: 2
        features: layering, striping, exclusive-lock, object-map, fast-diff, deep-flatten
        op_features: 
        flags: 
        create_timestamp: Tue May 17 04:43:48 2022
        access_timestamp: Tue May 17 04:43:48 2022
        modify_timestamp: Tue May 17 04:43:48 2022
        stripe unit: 4 KiB
        stripe count: 4
```

The change that the linked email talks about didn't affect `rbd_default_stripe_unit` and `rbd_default_stripe_count` handling at all.

> I can confirm that setting these options when creating an image does enable fancy striping, but creating an image and omitting them does not, even when they are specified in the rook-config-override configmap:

I'm not sure why it appeared to not work for @webberb with Rook.

**Comment by idryomov** [comment]
> @idryomov can we expose stripe-count and stripe-unit in the RBD storageclass and use them for creating RBD images? is there any caveat here?

No caveat -- that is how it should be done, similar to `imageFeatures`.  We really need to stop suggesting/using global config overrides for things that are not global at all.

**Comment by Madhu-1** [comment]
Thanks for confirming @idryomov

**Comment by Madhu-1** [comment]
@ajarr @kotreshhr what about cephfs? is it possible to provide these options during subvolume creation? or do we need to set them after mounting the subvolume?

**Comment by Madhu-1** [comment]
https://github.com/ceph/ceph-csi/issues/3124 to add support for rbd images.

**Comment by idryomov** [comment]
> But this would be a global setting instead of being customizable per volume or storage class.

Correct, but I don't think we should advertise this.  The right way to do it is through the storage class.

**Comment by travisn** [comment]
@idryomov Got it, so the support through the storage class would just require an update from the csi driver with https://github.com/ceph/ceph-csi/issues/3124.

**Comment by idryomov** [comment]
Yeah, the issue in Rook can be closed.

**Comment by travisn** [comment]
In this case I'm thinking we can leave this issue active to track when the csi driver update is integrated to rook.

**Comment by kahirokunn** [comment]
Keep open

**Comment by Madhu-1** [comment]
> @Madhu-1 What CSI release would this be included in? [ceph/ceph-csi#3132](https://github.com/ceph/ceph-csi/pull/3132)

@travisn it will be in the next cephcsi 3.7.0 release.

**Comment by travisn** [comment]
Ceph-CSI 3.7.0 is included with Rook v1.10 now. For an example of enabling striping, see [this rbd storage class](https://github.com/ceph/ceph-csi/blob/devel/examples/rbd/storageclass.yaml#L137-L144).

**Comment by Madhu-1** [comment]
Thanks @travisn Yes correct, we can close this one :+1:
